[WIP] GitHub integration

saundefined opened the pull request:
### Changed
- GitHub requests use a header instead of access_token

### Added
- Possibility to link GitHub profile

### SQL migration
```mysql
ALTER TABLE `users` ADD `github` VARCHAR(39) NULL AFTER `username`;
ALTER TABLE `users` ADD UNIQUE(`github`);
```

johannes commented on manage/users.php:
```diff
@@ -396,7 +419,8 @@ function csrf_validate(&$mydata, $name) {
 <?php if (!$unapproved) { ?>
-    <td colspan="2"><?php echo hsc($userdata['email']);?></td>
+    <td><?php echo hsc($userdata['email']);?></td>
+    <td><a href="https://github.com/<?php echo hsc($userdata['github']) ?>"><?php echo hsc($userdata['github']);?></a></td>
```
I haven't checked it, does `hsc()` use `ENT_QUOTES`? Otherwise the quotes for the attribute won't be escaped properly.

saundefined replied:
Yes, it uses.

```
function hsc($str){ return htmlspecialchars((string)$str,ENT_QUOTES,'UTF-8'); }
```

saundefined commented on manage/github.php:
```diff
@@ -83,15 +83,44 @@ function github_require_valid_user()
+    }
+    
+    $query = new Query('UPDATE users SET github = ? WHERE userid = ?', [$user->login, $id]);
+    db_query($query);
```
@johannes, thanks!

Added a check before the query exec, in case of an error during execution, an error is issued in `db_query()`

johannes replied:
Theoretically there is a chance for a race condition as there are no transactions, but I think for this use case here wa are fine. No escape of the username in the wanting output is tolerable here. For making life of Readers easier doingnit anyways might be useful, but no stopper.

saundefined commented on users.sql:
```diff
@@ -30,6 +30,7 @@ CREATE TABLE users (
   name varchar(255) NOT NULL default '',
   email varchar(255) NOT NULL default '',
   username varchar(16) default NULL,
+  github varchar(39) default NULL,
```
I'm not sure about the encoding, maybe Nikita will tell you.
39 characters — the maximum number of characters in a GitHub username.

nikic replied:
The encoding is utf8mb4 now.

johannes replied:
GitHub seems to limit to `[a-z0-9\-]` then we are fine. 

(Mind that "character" in unicode is a complicated topic, the grapheme cluster 👨‍👩‍👧‍👦 for instance has (if I count right) seven codepoints, thus won't fit in a `varchar(1)` (one codepoints for each person and zero width joiners in between))

johannes commented:
I haven't tested it, but looking over it seems fine.when applying this needs to be coordinated with schema changes and should be tested once master synced.

nikic commented:
This is currently using `scope=repo`. That makes sense to repo administration, but is way too much just to link accounts.

saundefined commented:
@nikic, 

> This is currently using `scope=repo`. That makes sense to repo administration, but is way too much just to link accounts.

I don't remove the repo administration section, if a user is in `GITHUB_PHP_OWNER_TEAM_ID`, it still can do it.
Also, token saved in the session, only if the user in the owners' team.

nikic commented:
> @nikic,
> 
> > This is currently using `scope=repo`. That makes sense to repo administration, but is way too much just to link accounts.
> 
> I don't remove the repo administration section, if a user is in `GITHUB_PHP_OWNER_TEAM_ID`, it still can do it.
> Also, token saved in the session, only if the user in the owners' team.

What I meant here is that the user shouldn't need to grant PHP write access to all their public and private repos just to associate their GitHub account. We should only request minimal permissions for this purpose.

saundefined commented:
> We should only request minimal permissions for this purpose.

replaced with `read:org`, without it I can't check membership

nikic reviewed (commented):
Having looked at what this code did before ... I think we can just drop the repo management part entirely. What it currently implements will not work right now (because the team it assigns doesn't exist anymore), and this doesn't really seem like particularly useful functionality in the first place. I wasn't aware this tool existed and have always created repositories without it...

saundefined commented on public/manage/github.php:
```diff
@@ -70,28 +79,75 @@ function github_require_valid_user()
+          '<p>Please contact an existing member if you see need.</p>';
+      foot();
+      exit;
+    }
```
@nikic,

I was thinking of a scenario like this:

we do not ask the user for additional scopes, just a name is enough.
Add a user with administrator rights to the PHP organization (e.g. @php-pulls or new).

When authorizing through GitHub, we link accounts and also using the admin user token, we check if there is a current user in the organization. If not yet, send an invite.

Is this workflow valid from an organizational security perspective?
Or is it preferable to leave control manually and through this script just link accounts?

nikic replied:
In the end, it will still be necessary to add the user to the correct team or grant repository permissions. An invite is automatically sent as part of doing that. As such, I don't think it's necessary to automate the sending of the invite here.

saundefined commented:
@nikic, thank you!

Updated the logic:
- link accounts in any case;
- don't request additional scopes;
- removed creation of repositories.

If it's ok, after #12, I'll update the changed functions and ready for final review.

Aaron-Junker commented:
Hi @saundefined. As you know I'm currently working on a RFC to create a global login system, that likely will use GitHub.

I just saw this PR and now I0m not sure if I should include this in the RFC and how, because it looks pretty finished, but it has no update since over a year.

saundefined commented:
@Aaron-Junker, this PR proposes simply linking GitHub profiles to a PHP account.
It should not affect global auth :)

Aaron-Junker commented:
> @Aaron-Junker, this PR proposes simply linking GitHub profiles to a PHP account. It should not affect global auth :)

Ok.
